### ISSUE TITLE
Add type hints to `pennylane.operation.Operator`

### ISSUE THREAD

**Issue by albi3ro** [PennyLaneAI/pennylane#5398]
[`Operator`](https://github.com/PennyLaneAI/pennylane/blob/78abf2b935ea3545f4e540d1da4b080a511b8a73/pennylane/operation.py#L406)  has a series of methods like `Operator.compute_matrix`, `Operator.matrix`, `Operator.compute_decomposition`, `Operator.compute_sparse_matrix` that raise errors by default, but return some kinda of representation when implemented.

Unfortunately, type checkers like Pylance then always interpret `Operator.compute_matrix` (and the others) to always return an error and consider all the code after such a method "unreachable":

<img width="680" alt="Screenshot 2024-03-18 at 9 40 26 AM" src="https://github.com/PennyLaneAI/pennylane/assets/6364575/53a2a0c5-75e2-498b-aec5-b52dba42c375">

This problem would be fixed if we added type hints to all the relevant `Operator` methods.  

This task would to be to add type-hints to all relevant Operator methods, for example:

```
    @staticmethod
    def compute_matrix(*params, **hyperparams) -> qml.typing.TensorLike: # pylint:disable=unused-argument
```

or:
```
    def terms(self) -> Tuple[List]:  # pylint: disable=no-self-use
```

Note the use of `qml.typing.TensorLike` for type hinting anything that could be an arbitrary style of numeric array.

The relevant docstring should sufficiently describe the expected type.

No test changes will be required for this change, but the contributor should add a changelog entry and add their name to the list of contributors.

**Comment by kenya-sk** [comment]
Hi, @albi3ro 

I'm interested in this issue. 
Can I try it as my first task?

**Comment by albi3ro** [comment]
Thanks for reaching out @kenya-sk .  I've assigned you the story. Feel free to reach out here or open a draft PR if you have any questions.

**Comment by kenya-sk** [comment]
@albi3ro 
I apologize for the delay in response. I am currently working on this task and could not reproduce the same error in my environment. could you please let me know if there are any conditions other than Pylance settings?


![スクリーンショット 2024-04-09 15 57 13](https://github.com/PennyLaneAI/pennylane/assets/30319295/6610b39f-cf24-4f57-af04-83fd490f6d3a)

**Comment by albi3ro** [comment]
One IDE-independent way to double-check this is by examining:

```
>>> qml.X(0).matrix.__annotations__
{'return': pennylane.typing.TensorLike}
>>> qml.X(0).sparse_matrix.__annotations__
{'return': scipy.sparse._csr.csr_matrix}
```

I would not recommend running `mypy` on PennyLane, as we have many, many type issues in need of fixing at the moment.

**Comment by albi3ro** [comment]
Closed by #5490
